### PR TITLE
[4.4.1] added GL problems detector and auto switch to software backend

### PR DESCRIPTION
**Pull request by igorkorsukov**
Resolves: https://github.com/musescore/MuseScore/issues/24269

### CODE REVIEW

File: src/framework/diagnostics/glproblemsdetector.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+{
+    GlobalConfiguration conf(modularity::globalCtx());
+    return conf.userAppDataPath().toQString() + "/need_use_software_quick_backend.dat";
+}
+
+void GlProblemsDetector::setIsNeedUseSoftwareRender(bool arg)
+{
+    if (arg) {
+        QFile file(softwareMarkerFilePath());
+        file.open(QIODevice::WriteOnly); // create file
```
Comment (cbjeukendrup) [inline src/framework/diagnostics/glproblemsdetector.cpp:90]: Need `file.close()` ?

**Review by RomanPudashkin — APPROVED** [review]

File: src/framework/diagnostics/glproblemsdetector.cpp [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    m_logDest->enabled = true;
+
+    Logger* logger = Logger::instance();
+    logger->addDest(m_logDest);
+}
+
+GlProblemsDetector::~GlProblemsDetector()
+{
+    //! NOTE log dest will deleted by logger
+    m_logDest->enabled = false;
```
Comment (cbjeukendrup) [inline src/framework/diagnostics/glproblemsdetector.cpp:77]: Can we maybe add `Logger::removeDest` and call that here? In order to have completely no influence on performance once the program is running

File: src/framework/diagnostics/glproblemsdetector.cpp (hunk 1 above)
Comment (cbjeukendrup) [reply src/framework/diagnostics/glproblemsdetector.cpp:90]: Never mind, it's done by `~QFile` of course

**Comment by cbjeukendrup** [comment]
Can we make it so that the version number is somehow included in the (name of) the file? Otherwise the software backend will continue to be used forever even if the problem is actually fixed in a next update.

File: src/framework/diagnostics/glproblemsdetector.cpp (hunk 2 above)
Comment (igorkorsukov) [reply src/framework/diagnostics/glproblemsdetector.cpp:77]: yes, that would be better...
I thought about it, but didn't dare...
I'll do it.

**Comment by igorkorsukov** [comment]
> Can we make it so that the version number is somehow included in the (name of) the file? Otherwise the software backend will continue to be used forever even if the problem is actually fixed in a next update.

good advice, I'll add

**Comment by mercuree** [comment]
Is there a possibility to force opengl? On musescore.org forum I suggested to use software backend because it is the first thing I [found](https://stackoverflow.com/questions/68154514/qt-qt-6-1-1-failed-to-create-vertex-shader-error-0x80070057) on Google and one musescore forum user confirmed that it works. 
But personally software backend does not work perfectly for me. For example, thumbnails are not displayed in New & Recent window.
I have Intel Graphics UHD 630 and it works fine without any workarounds.

**Comment by igorkorsukov** [comment]
> Is there a possibility to force opengl? On musescore.org forum I suggested to use software backend because it is the first thing I [found](https://stackoverflow.com/questions/68154514/qt-qt-6-1-1-failed-to-create-vertex-shader-error-0x80070057) on Google and one musescore forum user confirmed that it works. But personally software backend does not work perfectly for me. For example, thumbnails are not displayed in New & Recent window. I have Intel Graphics UHD 630 and it works fine without any workarounds.

We don't force the software backend on for everyone here.
We check, if we find a problem with opengl, then we switch to software, otherwise opengl will remain as is.

File: src/app/internal/guiapp.cpp [hunk 3]
```diff
@@ -149,8 +164,21 @@ void GuiApp::perform()
     const QUrl url(QStringLiteral("qrc:/qml") + mainQmlFile);
 #endif
 
-    QObject::connect(engine, &QQmlApplicationEngine::objectCreated,
-                     qApp, [this, url, splashScreen](QObject* obj, const QUrl& objUrl) {
+    QObject::connect(engine, &QQmlApplicationEngine::objectCreated, qApp,
+                     [this, url, splashScreen, glProblemsDetector](QObject* obj, const QUrl& objUrl) {
+        if (glProblemsDetector->isProblemWithGl()) {
```
Comment (cbjeukendrup) [inline src/app/internal/guiapp.cpp:169]: It seems that if this condition is false, the `glProblemsDetector` will not be deleted

Comment (igorkorsukov) [reply src/app/internal/guiapp.cpp:169]: fixed

**Comment by igorkorsukov** [comment]
> @igorkorsukov But now we are assuming that OpenGL is the problem, and we fall back to Software. But what if it's completely different, and Direct3D is the problem and we should fall back to OpenGL (or still Software)?
> 
> I was looking at https://bugreports.qt.io/browse/QTBUG-100689, which is a duplicate of https://bugreports.qt.io/browse/QTBUG-78648, which lists several related patches and they all have to do with d3d (which means Direct3D if I'm not mistaken).

It would be possible to experiment with this, and I also thought that it would be possible to try several backends...
But: 
* We, the developers, don't have this reproducible issue, so it will be difficult for us to test and make sure that we are doing everything correctly.
* As I understand it, the problem is quite critical and should be solved quickly... Now we have done it.
* The current solution is quite simple and then quite reliable, while the solution of iterating over different backends is more complex and therefore less reliable. 


In general, I suggest accepting this fix now. And continue investigate the problem... Maybe add checking for multiple backends to the next version, and also remove opengl_buglist, it will no longer be needed, and also add information about the backend to the diagnostics, etc.

**Review by cbjeukendrup — APPROVED** [review]
Okay, let's go with this solution for 4.4.1 then. On master let's continue investigating, and try updating to Qt 6.7 or Qt 6.8 (the bug should be fixed in Qt 6.4).

(the only thing one could say is that it is now maybe incorrectly named "*Gl*ProblemsDetector" because OpenGL might not be involved, but let's not change that now, not worth the effort)

**Comment by igorkorsukov** [comment]
> Okay, let's go with this solution for 4.4.1 then. On master let's continue investigating, and try updating to Qt 6.7 or Qt 6.8 (the bug should be fixed in Qt 6.4).
> 
> (the only thing one could say is that it is now maybe incorrectly named "_Gl_ProblemsDetector" because OpenGL might not be involved, but let's not change that now, not worth the effort)

GL - means only graphic layer :)

**Pull request merged**